### ISSUE TITLE
ObjectifyJS

### ISSUE THREAD

**Issue by ersimont** [objectify/objectify#346]
Hello!

I used and loved Objectify for a project of mine starting with version 3. Now I'm working with a project in Node, and I miss it! I see [Pebblebed](https://github.com/lostpebble/pebblebed) says to be "heavily inspired" by this library, and [gstore-node](https://github.com/sebelga/gstore-node) is another JS library with a similar purpose, but both lack a main feature I desire: declaring my "schema" with decorators, which in my mind is part of the essence of Objectify.

I've started playing with creating such a feature, and my POC is working so far. Using Typescript I have code like this functional:

```ts
@Entity()
class Project {
  @Id() id: string;
  @Index() name: string;
  embedded: SomeOtherClass[];
  @Ignore() ignored: boolean;
}

ofy.save().entity(someProject);
ofy.save().entities(project1, project2);

ofy.load().entity(someProject);
ofy.load().key(keyOf(someProject));
ofy.load().type(Project).id(1);
ofy.load().kind('Project').ids(1, 2, 3);
```

I don't know for sure that I'll have the drive to push this through to a polished library, but in case I do:
- Do you happen to know of another library I've missed that I should be using, instead of writing my own?
- Do you like or dislike the idea of a "clone" library in another language that clearly is just trying to imitate your exact API?
- Do you have any advice or ideas for such an undertaking - e.g. things you would change if you were starting over, or words of warning/encouragement?

Thank you for a great library, and the countless hours you put into making it great for the rest of us!

**Comment by ersimont** [comment]
Ah - I see there are already a few changes listed in the [v6 design](https://github.com/objectify/objectify/wiki/DesignOfObjectify6)!

How stable is the code in the V6 branch? Would it make sense for me to start modeling my code off what is in that branch?

**Comment by stickfigure** [comment]
I'm glad to have inspired other projects! And I'll find somewhere to link to your project when you publish a release. I'm amazed Typescript has that kind of runtime type info available. Very cool.

Yep - v6 has been merged into master and the first alpha release has been cut. Hoping to get into beta in Jan. Follow master.

I don't know what advice to offer you, but I'm happy to respond to questions - my email address is jeff infohazard org with the predictable at-sign location. Or ask on the objectify google group. I suspect a literal translation to Node.js will be hard since Objectify is built around Futures, which are quite different from Node's promises/callbacks. There's also a _lot_ of code here! I guess my main advice would be the usual "get something simple working first and add features over time".

There are probably lots of things I would do differently if I started from scratch, but most of the egregious issues have been addressed over the years. Of course that evolution created some cruft too - sorry about that. There's some really ugly logic setting up polymorphic classtranslators that would be a lot easier if "registration" had some sort of done event (ie, all registration happens in a closure). Honestly, doing the whole thing with callbacks is such a different beast that it's hard for me to wrap my head around it without prototyping.
